### PR TITLE
Enable boolean indices for SparseList

### PR DESCRIPTION
**Pull request by pmrv**
This makes using it for selective dynamics a bit more ergonomic, since you can do something like
```python
structure.selective_dynamics[structure.get_chemical_elements()=='H'] = [True, False, False]
```
directly without calculating the indices yourself.

### CODE REVIEW

File: pyiron_atomistics/atomistics/structure/sparse_list.py [hunk 1]
```diff
@@ -183,13 +183,16 @@ def __getitem__(self, item):
...
-                elif isinstance(item[0], (bool, np.bool_)):
+                # for some reason isinstance(True, int) == True, so check for bools first explicitly
+                if isinstance(item[0], (bool, np.bool_)):
+                    if len(item) != len(self):
+                        raise IndexError("Length of boolean index does not match indexed list!")
                     ind_list = []
                     for i, bo in enumerate(item):
                         if bo:
                             ind_list.append(i)
+                elif isinstance(item[0], (int, np.integer)):
```
Comment (sudarsan-surendralal) [inline pyiron_atomistics/atomistics/structure/sparse_list.py]: Should it also be `np.int_`? (I'm not sure though)

File: pyiron_atomistics/atomistics/structure/sparse_list.py [hunk 2]
```diff
@@ -183,13 +183,16 @@ def __getitem__(self, item):
...
-                    ind_list = item
-                elif isinstance(item[0], (bool, np.bool_)):
+                # for some reason isinstance(True, int) == True, so check for bools first explicitly
+                if isinstance(item[0], (bool, np.bool_)):
+                    if len(item) != len(self):
+                        raise IndexError("Length of boolean index does not match indexed list!")
                     ind_list = []
                     for i, bo in enumerate(item):
                         if bo:
                             ind_list.append(i)
```
Comment (sudarsan-surendralal) [inline pyiron_atomistics/atomistics/structure/sparse_list.py]: Not related to what you wrote but:

```suggestion
                    ind_list = np.argwhere(item).flatten()
```

File: pyiron_atomistics/atomistics/structure/sparse_list.py [hunk 3]
```diff
@@ -209,8 +212,15 @@ def __setitem__(self, key, value):
...
         if max(key) > self._length:
             raise IndexError
-        for i in key:
-            self._dict[i] = value
+
+        keys = list(key)
+        if isinstance(keys[0], (bool, np.bool_)):
+            for i, bo in enumerate(keys):
+                if bo:
+                    self._dict[i] = value
```
Comment (sudarsan-surendralal) [inline pyiron_atomistics/atomistics/structure/sparse_list.py]: See suggestion above (using `np.argwhere`)

File: tests/atomistics/structure/test_sparse_list.py [hunk 4]
```diff
@@ -44,6 +44,16 @@ def test__getitem__(self):
...
+        self.assertEqual(self.aList[[False, True, False, False, True, False, False]].list(),
+                         [True, True],
+                         "Indexing with boolean mask returns wrong values!")
+        cListModified = self.cList.__copy__()
+        cListModified[[True, True, False, False, False]] = 0.1
+        self.assertEqual(cListModified[0], 0.1,
+                         "Indexing with boolean mask sets wrong values!")
+        self.assertEqual(cListModified[1], 0.1,
+                         "Indexing with boolean mask sets wrong values!")
+
```
Comment (sudarsan-surendralal) [inline tests/atomistics/structure/test_sparse_list.py:62]: Nitpick: For completeness why not check that any/all of the other indices aren't 0.1

**Review by sudarsan-surendralal — COMMENTED** [review]
Thanks, this is a necessary feature. A few nitpicks:

File: pyiron_atomistics/atomistics/structure/sparse_list.py (hunk 1 above)
Comment (pmrv) [reply pyiron_atomistics/atomistics/structure/sparse_list.py]: `np.int_` is a shorthand for the native int for the current platform, whereas `np.integer` really catches all numpy integer types.  I should be using `numbers.Integral` though, I guess.

**Review by sudarsan-surendralal — APPROVED** [review]
LGTM!

**Pull request merged**